src/D3MHub.sol

```diff
@@ -221,121 +219,118 @@ contract D3MHub {
-        } else if (mode == Mode.MCD_CAGED) {
-            // MCD caged
-            // debt is obtained from free collateral owned by the End module
-            _end = end;
-            _end.skim(ilk, address(_pool));
-            daiDebt = vat.gem(ilk, address(_end));
-        } else revert("D3MHub/unknown-mode");
-
-        uint256 availableAssets = _pool.maxWithdraw();
-
-        // Unwind amount is limited by how much:
-        // - max reduction desired
-        // - assets available
-        // - dai debt tracked in vat (CDP or free)
-        uint256 amount = _min(
-                            _min(
-                                supplyReduction,
-                                availableAssets
-                            ),
-                            daiDebt
-                        );
-        require(amount <= MAXINT256, "D3MHub/overflow");
+    function _unwind(bytes32 ilk, ID3MPool _pool, uint256 amount) internal {
+        if (amount > 0) {
+            _pool.withdraw(amount);
+            daiJoin.join(address(this), amount);
+            vat.frob(ilk, address(_pool), address(_pool), address(this), -int256(amount), -int256(amount));
+            vat.slip(ilk, address(_pool), -int256(amount));
+        }
+        emit Unwind(ilk, amount);
+    }
 
-        // Determine the amount of fees to bring back
-        uint256 fees = 0;
-        if (assetBalance > daiDebt) {
-            unchecked {
-                fees = assetBalance - daiDebt;
-            }
+    function _unwind(bytes32 ilk, ID3MPool _pool, address urn) internal {
+        uint256 amount = _pool.maxWithdraw();
+        if (amount > 0) {
+            uint256 toSlip = _min(vat.gem(ilk, urn), amount);
+            require(toSlip <= MAXINT256, "D3MHub/overflow");
+            vat.slip(ilk, urn, -int256(toSlip));
 
-            if (amount + fees > availableAssets) {
-                unchecked {
-                    fees = availableAssets - amount;
-                }
-            }
+            _pool.withdraw(amount);
+            daiJoin.join(address(this), amount);
+            vat.move(address(this), vow, amount * RAY);
         }
+        emit Unwind(ilk, amount);
+    }
 
-        if (amount == 0 && fees == 0) {
-            emit Unwind(ilk, 0, 0);
-            return;
+    function _normal(bytes32 ilk, ID3MPool _pool, uint256 Art, uint256 lineWad) internal {
+        (uint256 ink, uint256 art) = vat.urns(ilk, address(_pool));
+        require(art == Art, "D3MHub/more-than-one-urn");
+        uint256 currentAssets = _pool.assetBalance(); // Should return DAI owned by D3MPool
+        uint256 maxWithdraw = _pool.maxWithdraw();
+        if (currentAssets > ink) { // If fees were generated
+            uint256 fixInk = _min(
+                currentAssets - ink, // fees generated
+                ink < lineWad // if previously CDP was under debt ceiling
+                    ? (lineWad - ink) + maxWithdraw // up to gap to reach debt ceiling + maxWithdraw
```

Comment:
if we have set `Line < line` then we could end up in a state where we add `ink` above the Global Debt Ceiling and cannot unwind enough at the end.  We probably want to have a check like `debt > Line ? _divup(debt - Line, RAY) : 0 // global debt ceiling exceeded` here too so `lineWad` becomes either the `Line` or the `line` depending on which is less.

Reply:
we might even be able to move this to `exec` before normal and do something like:

```solidity
uint256 Line = vat.Line();
uint256 lineWad = min(line, Line) / RAY;
_normal(
  ilk,
  _pool,
  Art,
  lineWad
);
```

Reply:
I think this is a different thing. Any `ilk` can go above local debt ceiling or global debt ceiling due fees generation. There is not a restriction for this.
In this case we added this restriction just a safety precaution that something is going wrong with the fee calculation (as we are outsourcing it to a third party).
I think in this case, using the local debt ceiling is more than enough for this purpose as otherwise it gets more complex.
Also note that for knowing if you are passing the general debt ceiling you actually need to calculate the gap between `debt` and `Line` minus what you already have in use.

Reply:
I mean you can't do: `min(line, Line)` as `Line` is not exclusive for this CDP as `line` is.